### DISCUSSION TITLE
Building CapnProto

### DISCUSSION THREAD

**Discussion by kentdlee** [capnproto/capnproto#1932] [General]
I am working on an open source project (dragonhpc.org) and building capnproto to be included in our project. I am having a few issues that I am hoping someone can help with.

First, when I build the repo (master branch) I get the following error only when compiling on one of our internal test machines. I don't get this error when compiling in a container. 

```
1052 test(s) passed
1 test(s) failed
FAIL: capnp-test
Randomly testing backwards-compatibility scenarios with seed: 1706979572
PASS: capnp-evolution-test
PASS: src/capnp/compiler/capnp-test.sh
===========================================
1 of 3 tests failed
Please report to capnproto@googlegroups.com
===========================================
```

I didn't include all of the text because there is a lot printed of course when running tests. If I should gather some other error information and report it somewhere, please let me know. 

The other issue I have is knowing how best to use capnproto in the project without installing it on the internal build machine. So far I have been reaching into the capnproto/c++/.libs directory to find the libraries (capnp and kj) and into the capnproto/c++/src directory to find the includes. I admit I am not an expert on makefiles, so asking for some help in knowing what the best practice would be here. 

Thanks for anything you can help me with.

**Comment by kentdlee** [comment]
I should have added that I am building on a standard Ubuntu machine.

**Comment by kentonv** [comment]
I would need to see the test log, in particular for the failing test, in order to say what went wrong here.

Usually if you don't want to install globally on the machine, you'd use `./configure --prefix=/some/path` to arrange to install to `/some/path` instead, which can be somewhere inside your build directory.

Note that you might also benefit from `./configure --disable-shared`. This will disable building shared libraries and instead only build static libraries. You may find these are easier to work with as you don't have to dig into the `.libs` directory and you don't have to figure out how to make the libraries findable at program startup time, only at link time.

**Reply by kentdlee** [reply]
Kenton, thanks for the pointers. I appreciate it. Can you point me to the test log file so I can provide that to you?

The prefix path is a good suggestion. I appreciate that.

I did try over the weekend to use the .a libraries and statically link them into our code. We are using a combination of gcc and g++. Not sure that has anything to do with the issue, but when I try to link with the non-shared .a files I got an error indicating that it was not position independent code and to use -fPIC. I can see that capnproto is compiled that way so I was a little lost at that point as to how to make that work. The .so files are working, but would rather statically link code into our .so. 

Again, thanks for any help/direction you can provide.

  -- Kent

**Reply by kentdlee** [reply]
I did find an errors.txt file in capnproto/c++/src/capnp/testdata. Here are the contents if that's the correct log file. 

file:74:30-32: error: As of Cap'n Proto v0.3, it is no longer necessary to assign numbers to unions. However, removing the number will break binary compatibility. If this is an old protocol and you need to retain compatibility, please add an exclamation point after the number to indicate that it is really needed, e.g. `foo @1! :union {`. If this is a new protocol or compatibility doesn't matter, just remove the @n entirely. Sorry for the inconvenience, and thanks for being an early adopter!  :)
file:74:30-32: error: As of Cap'n Proto v0.3, the 'union' keyword should be prefixed with a colon for named unions, e.g. `foo :union {`.
file:79:23-25: error: As of Cap'n Proto v0.3, it is no longer necessary to assign numbers to unions. However, removing the number will break binary compatibility. If this is an old protocol and you need to retain compatibility, please add an exclamation point after the number to indicate that it is really needed, e.g. `foo @1! :union {`. If this is a new protocol or compatibility doesn't matter, just remove the @n entirely. Sorry for the inconvenience, and thanks for being an early adopter!  :)
file:84:17-19: error: As of Cap'n Proto v0.3, the 'union' keyword should be prefixed with a colon for named unions, e.g. `foo :union {`.
file:133:7-10: error: 'using' declaration without '=' must specify a named declaration from a different scope.
file:37:3-10: error: 'dupName' is already defined in this scope.
file:36:3-10: error: 'dupName' previously defined here.
file:52:5-12: error: 'dupName' is already defined in this scope.
file:36:3-10: error: 'dupName' previously defined here.
file:55:3-8: error: An unnamed union is already defined in this scope.
file:51:3-8: error: Previously defined here.
file:60:10-24: error: Type names must begin with a capital letter.
file:61:3-18: error: Non-type names must begin with a lower-case letter.
file:62:3-14: error: Cap'n Proto declaration names should use camelCase and must not contain underscores. (Code generators may convert names to the appropriate style for the target language.)
file:66:5-27: error: This kind of declaration doesn't belong here.
file:44:3-23: error: Union must have at least two members.
file:45:3-23: error: Group must have at least one member.
file:47: error: Union must have at least two members.
file:92: error: Unions cannot contain unnamed unions.
file:39:15-16: error: Duplicate ordinal number.
file:38:15-16: error: Ordinal @2 originally used here.
file:41:18-19: error: Skipped ordinal @3.  Ordinals must be sequential with no holes.
file:69:15-17: error: Union ordinal, if specified, must be greater than no more than one of its member ordinals (i.e. there can only be one field retroactively unionized).
file:117:31-50: error: Import failed: noshuchfile.capnp
file:119:26-32: error: Not defined: NoSuch
file:120:28-34: error: 'Foo' has no member named 'NoSuch'
file:97:25-29: error: 'List' requires exactly one parameter.
file:98:30-48: error: Too many generic parameters.
file:98:30-34: error: 'List' requires exactly one parameter.
file:99:23-39: error: 'List(AnyPointer)' is not supported.
file:101:17-24: error: 'notType' is not a type.
file:102:17-27: error: Declaration does not accept generic parameters.
file:104:34-41: error: Integer value out of range.
file:105:37-38: error: Integer value out of range.
file:106:32-35: error: Type mismatch; expected Text.
file:107:33-38: error: Type mismatch; expected Text.
file:108:33-55: error: Type mismatch; expected Text.
file:109:43-61: error: Integer is too big to be negative.
file:110:35-39: error: '.Foo' does not refer to a constant.
file:111:44-51: error: Constant names must be qualified to avoid confusion.  Please replace 'notType' with '.notType', if that's what you intended.
file:118:28-34: error: Not defined: NoSuch
file:100:22-37: error: 'List(AnyStruct)' is not supported.
file:113:29-32: error: 'Foo' is not an annotation.
file:114:29-47: error: 'notFieldAnnotation' cannot be applied to this kind of declaration.
file:115:33-48: error: 'fieldAnnotation' requires a value.
file:127:35-46: error: Struct has no field named 'nosuchfield'.
file:128:49-52: error: Type mismatch; expected group.
file:126:52-55: error: Missing field name.
file:137:3-10: error: 'dupName' is already defined in this scope.
file:136:3-10: error: 'dupName' previously defined here.
file:139:15-16: error: Duplicate ordinal number.
file:138:15-16: error: Ordinal @2 originally used here.
file:142:7-16: error: Declaration recursively depends on itself.
file:145:19-20: error: 'T' is not an annotation.
file:149:14-27: error: Not enough generic parameters.
file:150:15-47: error: Too many generic parameters.
file:151:18-49: error: Double-application of generic parameters.
file:152:38-43: error: Sorry, only pointer types can be used as generic parameters.
file:155:30-44: error: Embeds can only be used when Text, Data, or a struct is expected.
file:156:37-51: error: Couldn't read file for embed: no-such-file
file:162:23-27: error: Only pointer parameters can declare their default as 'null'.
file:163:10-16: error: 'stream' can only appear after '->', not before.
file:163:10-16: error: A method declaration uses streaming, but '/capnp/stream.capnp' is not found in the import path. This is a standard file that should always be installed with the Cap'n Proto compiler.
file:158:20-45: error: Import failed: nosuchfile-unused.capnp

**Reply by kentdlee** [reply]
OK, some further information. This is the test that fails.

[ FAIL ] kj/async-io-test.c++:309: legacy test: AsyncIo/AncillaryMessageHandler (630 μs)

**Reply by kentdlee** [reply]
g++ and gcc are at 11.2.0 if that is helpful.

**Reply by kentonv** [reply]
The errors.txt file whose contents you pasted here is [checked into the repository](https://github.com/capnproto/capnproto/blob/v2/c%2B%2B/src/capnp/testdata/errors.txt) -- it doesn't contain errors generated from your particular test run. Rather, it's part of a test that checks that the compiler reports errors correctly.

The relevant errors explaining your test failure would appear immediately before the `[ FAIL ]` line you copy-pasted. FWIW, that particular test is known to fail on Linux kernel versions before 4.14, so if you're using an ancient kernel that could be the issue. In any case, the test doesn't really matter, nothing in Cap'n Proto actually uses that functionality, so you can ignore it.

By default, automake will compile every source file twice, once for static libraries (`.a`) and once for dynamic (`.so`). Only the dynamic libraries are compiled with `-fPIC`, so that explains what you're seeing. This is just the traditional way automake works, it's the same for all projects that use it. You could probably fiddle with `configure` flags to make it compile the static libs with `-fPIC`, but this isn't something I'm deeply familiar with.

**Reply by kentdlee** [reply]
Thanks for the late at night reply. I appreciate the information.

The kernel is indeed before 4.14 (4.12) so that's the issue. 

I don't need rpc at all so that shouldn't really be an issue for me. 

I'll explore the configure flags to see if I can't get the fPIC to work. 

Thank you! I appreciate it.

**Reply by kentdlee** [reply]
I guess I have one more question. How can I get the make to ignore that error and continue the build? It errors out after that error. I don't need the RPC functionality. Is there a way to just configure without it?

**Reply by kentonv** [reply]
That's the end of the build. There's nothing else to continue to. If you'd rather not run the tests at all, you can run plain `make` instead of `make check`.

The test failure does not mean RPC is broken. The test itself tries to use a feature that was new in 4.14, but none of Cap'n Proto (RPC or otherwise) actually uses that feature.

**Reply by kentdlee** [reply]
Thanks. That helps greatly. I appreciate it.
